### PR TITLE
chore: changed generate tiles algo (MAPCO-2807)

### PR DESCRIPTION
**Pull request by CL-SHLOMIKONCHA**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✖                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✔                                                                        |
| Chore            | ✔                                                                       |

### CODE REVIEW

File: src/createPackage/models/createPackageManager.ts [hunk 1]
```diff
@@ -88,7 +87,7 @@ export class CreatePackageManager {
     const srcRes = layerMetadata.maxResolutionDeg as number;
     const maxZoom = degreesPerPixelToZoomLevel(srcRes);
     if (zoomLevel > maxZoom) {
-      throw new BadRequestError(`The requested requested resolution ${targetResolution} is larger then then product resolution ${srcRes}`);
+      throw new BadRequestError(`The requested requested resolution ${targetResolution} is larger than product resolution ${srcRes}`);
```
Comment (asafMasa) [inline src/createPackage/models/createPackageManager.ts]: 👍

File: src/createPackage/models/createPackageManager.ts [hunk 2]
```diff
@@ -160,7 +155,7 @@ export class CreatePackageManager {
       cswProductId: resourceId,
       crs: crs ?? DEFAULT_CRS,
       productType,
-      batches,
+      batches: batches as ITileRange[],
```
Comment (asafMasa) [inline src/createPackage/models/createPackageManager.ts]: why do you need the cast?

File: src/createPackage/models/createPackageManager.ts [hunk 3]
```diff
@@ -267,6 +262,24 @@ export class CreatePackageManager {
     return sanitized;
   }
 
+  private generateTileGroups(polygon: Polygon, footprint: Polygon | MultiPolygon, zoom: number): ITileRange[] | null {
+    const intersaction = intersect(polygon, footprint);
```
Comment (asafMasa) [inline src/createPackage/models/createPackageManager.ts]: protect with try catch
if error print error details (requested polygon and footprint)

File: src/createPackage/models/createPackageManager.ts [hunk 4]
```diff
@@ -267,6 +262,24 @@ export class CreatePackageManager {
...
+    const tilesGroups: ITileRange[] = [];
+
+    if (intersaction === null) {
+      return null;
+    }
+
+    for (let i = 0; i <= zoom; i++) {
+      const zoomTilesGroups = new TileRanger().encodeFootprint(intersaction, i);
```
Comment (asafMasa) [inline src/createPackage/models/createPackageManager.ts]: protect with try catch
if error print error details

File: src/createPackage/models/createPackageManager.ts [hunk 5]
```diff
@@ -267,6 +262,24 @@ export class CreatePackageManager {
...
 
+  private generateTileGroups(polygon: Polygon, footprint: Polygon | MultiPolygon, zoom: number): ITileRange[] | null {
+    const intersaction = intersect(polygon, footprint);
+    const tilesGroups: ITileRange[] = [];
+
+    if (intersaction === null) {
+      return null;
+    }
+
+    for (let i = 0; i <= zoom; i++) {
```
Comment (asafMasa) [inline src/createPackage/models/createPackageManager.ts]: you dont need the for, just encode with max zoom level and check the result GPKG

**Review by asafMasa — CHANGES_REQUESTED** [review]
👍 Good job

File: src/createPackage/models/createPackageManager.ts (hunk 3 above)
Comment (CL-SHLOMIKONCHA) [reply src/createPackage/models/createPackageManager.ts]: imo, we can only check if intersaction is null and throw new BadRequestError with those details as we did to the sanatized (uses the same intersaction func from turf)

File: src/createPackage/models/createPackageManager.ts (hunk 5 above)
Comment (CL-SHLOMIKONCHA) [reply src/createPackage/models/createPackageManager.ts]: false, as the encode function expect for the max zoom and the batches output will be include only the requested zoom (we must go trough all zoom levels)

File: src/createPackage/models/createPackageManager.ts (hunk 4 above)
Comment (CL-SHLOMIKONCHA) [reply src/createPackage/models/createPackageManager.ts]: same as the previous comment - we can only check if intersaction is null and throw new BadRequestError with those details as we did to the sanatized (uses the same intersaction func from turf)

File: src/createPackage/models/createPackageManager.ts (hunk 2 above)
Comment (CL-SHLOMIKONCHA) [reply src/createPackage/models/createPackageManager.ts]: fixed

File: src/createPackage/models/createPackageManager.ts (hunk 3 above)
Comment (asafMasa) [reply src/createPackage/models/createPackageManager.ts]: https://stackoverflow.com/questions/61926747/topologyexception-when-using-turf-intersect
as you can see turf.intersect can throw errors, so i prefer we will add try..catch when using turf functions for better error handling, if we don't do it - users may get 500 without knowing why the package failed

File: src/createPackage/models/createPackageManager.ts (hunk 4 above)
Comment (asafMasa) [reply src/createPackage/models/createPackageManager.ts]: same as before, as it uses turf inside i prefer you'll protect our code with try..catch for better error handling

File: src/createPackage/models/createPackageManager.ts (hunk 5 above)
Comment (asafMasa) [reply src/createPackage/models/createPackageManager.ts]: Open an issue regarding the expected and actual behavior of the algorithm and i'll approve this

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: src/createPackage/models/createPackageManager.ts (hunk 3 above)
Comment (CL-SHLOMIKONCHA) [reply src/createPackage/models/createPackageManager.ts]: fixed

File: src/createPackage/models/createPackageManager.ts (hunk 4 above)
Comment (CL-SHLOMIKONCHA) [reply src/createPackage/models/createPackageManager.ts]: fixed

File: src/createPackage/models/createPackageManager.ts (hunk 5 above)
Comment (CL-SHLOMIKONCHA) [reply src/createPackage/models/createPackageManager.ts]: done [](https://mapcolonies.atlassian.net/browse/MAPCO-2871)

Comment (CL-SHLOMIKONCHA) [reply src/createPackage/models/createPackageManager.ts]: https://mapcolonies.atlassian.net/browse/MAPCO-2871

File: src/createPackage/models/createPackageManager.ts [hunk 6]
```diff
@@ -259,12 +254,39 @@ export class CreatePackageManager {
...
+      }
+
+      for (let i = 0; i <= zoom; i++) {
+        const zoomTilesGroups = new TileRanger().encodeFootprint(intersaction, i);
+        for (const group of zoomTilesGroups) {
+          tilesGroups.push(group);
+        }
+      }
+      return tilesGroups;
+    } catch (error) {
```
Comment (asafMasa) [inline src/createPackage/models/createPackageManager.ts:301]: this catch will catch the 400 bad request exception...
separate the 400 catch from the encodeFootprint try catch

**Review by asafMasa — CHANGES_REQUESTED** [review]

**Review by asafMasa — APPROVED** [review]

**Pull request merged**